File: app/src/main/java/space/itoncek/trailcompass/app/hideandseek/ui/LoginActivity.java
package space.itoncek.trailcompass.app.hideandseek.ui;

import static space.itoncek.trailcompass.client.api.LoginResponse.UNABLE_TO_CONNECT;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import org.json.JSONException;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.MessageFormat;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import space.itoncek.trailcompass.app.R;
import space.itoncek.trailcompass.app.hideandseek.api.HideAndSeekAPIFactory;
import space.itoncek.trailcompass.client.api.HideAndSeekAPI;
import space.itoncek.trailcompass.client.api.HideAndSeekConfig;
import space.itoncek.trailcompass.client.api.LoginResponse;

public class LoginActivity extends AppCompatActivity {
    ProgressBar mProgressBar;
    TextView mProgressLabel;
    HideAndSeekAPI api;

    File mapfile;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_hideandseek_login);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        mapfile = new File(getCacheDir() + "/map.map");

        mProgressBar = findViewById(R.id.loading_progressbar);
        mProgressLabel = findViewById(R.id.loading_label);

        api = new HideAndSeekAPIFactory(getFilesDir()).generateHideAndSeekAPI();

        CountDownLatch cdl = new CountDownLatch(1);
        AtomicReference<LoginResponse> login = new AtomicReference<>();
        Thread t = new Thread(() -> {
            runOnUiThread(()->mProgressLabel.setText(String.format(getString(R.string.login_started) , api.getConfig().username)));
            try {
                login.set(api.login());
            } catch (IOException | JSONException e) {
                Log.e(LoginActivity.class.getName(), getResources().getText(R.string.login_unable_to_connect).toString(), e);
                Toast.makeText(this, getResources().getText(R.string.login_unable_to_connect), Toast.LENGTH_LONG).show();
                login.set(UNABLE_TO_CONNECT);
            }
            cdl.countDown();
        });
        t.start();

        try {
            cdl.await();
        } catch (InterruptedException e) {
            Log.d(LoginActivity.class.getName(), "Interrupted", e);
        }

        Toast.makeText(this, login.get().name(), Toast.LENGTH_SHORT).show();
        switch (login.get()) {
            case UNABLE_TO_CONNECT -> {
                runOnUiThread(()->mProgressLabel.setText(R.string.login_unable_to_connect));
                Toast.makeText(this, getResources().getText(R.string.login_unable_to_connect), Toast.LENGTH_LONG).show();
                this.finish();
            }
            case INCOMPATIBLE_BACKEND -> {
                runOnUiThread(()->mProgressLabel.setText(R.string.login_wrong_version));
                Toast.makeText(this, getResources().getText(R.string.login_wrong_version), Toast.LENGTH_LONG).show();
                this.finish();
            }
            case UNABLE_TO_AUTH -> {
                runOnUiThread(()->mProgressLabel.setText(R.string.login_wrong_password));
                Toast.makeText(this, getResources().getText(R.string.login_wrong_password), Toast.LENGTH_LONG).show();
                this.finish();
            }
            case OK -> {
                runOnUiThread(()->mProgressLabel.setText(String.format(getString(R.string.login_succes),api.getConfig().username)));
                downloadMapUpdates();
            }
        }

    }

    private void downloadMapUpdates() {
        // execute this when the downloader must be fired
        Thread t = new Thread(() -> {
            AtomicBoolean isCanceled = new AtomicBoolean(false);
            findViewById(R.id.loading_cancel).setOnClickListener((c) -> isCanceled.set(true));
            try {
                HideAndSeekConfig c = api.getConfig();

                runOnUiThread(()->mProgressLabel.setText(R.string.login_checking_map_hash));
                String str = api.getMapHash();

                if (!str.equals(c.map_hash) || !mapfile.exists()) {
                    c.map_hash = str;
                    api.saveConfig(c);
                    runOnUiThread(()->mProgressLabel.setText(R.string.login_map_update_needed));
                    InputStream input = null;
                    OutputStream output = null;
                    HttpURLConnection connection = null;
                    try {
                        URL url = new URL(c.base_url + "/mapserver/getServerMap");
                        connection = (HttpURLConnection) url.openConnection();
                        connection.setRequestProperty("Authorization","Bearer " + c.jwt_token);
                        connection.connect();

                        // expect HTTP 200 OK, so we don't mistakenly save error report
                        // instead of the file
                        if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                            throw new RuntimeException(new IOException("Server returned HTTP " + connection.getResponseCode()
                                    + " " + connection.getResponseMessage()));
                        }

                        // this will be useful to display download percentage
                        // might be -1: server did not report the length
                        int fileLength = connection.getContentLength();
                        Log.i("LoginActivity", "FL " + fileLength);
                        // download the file
                        input = connection.getInputStream();
                        output = new FileOutputStream(mapfile);

                        byte[] data = new byte[65535];
                        long total = 0;
                        int count;
                        while ((count = input.read(data)) != -1) {
                            // allow canceling with back button
                            if (isCanceled.get()) {
                                input.close();
                            }
                            total += count;
                            // publishing the progress....
                            if (fileLength > 0) // only if total length is known
                            {
                                int finalTotal = (int) total;
                                runOnUiThread(()-> {
                                    mProgressBar.setIndeterminate(false);
                                    mProgressBar.setProgress((int) (100*((float)finalTotal/fileLength)),true);
                                    mProgressBar.setMax(100);

                                    mProgressLabel.setText(MessageFormat.format(getString(R.string.login_map_download_progress), (int) (100 * ((float) finalTotal / fileLength))));
                                });
                            }
                            output.write(data, 0, count);
                        }
                        Log.i("LoginActivity","total "  + total);
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    } finally {
                        try {
                            if (output != null)
                                output.close();
                            if (input != null)
                                input.close();
                        } catch (IOException ignored) {
                        }

                        if (connection != null)
                            connection.disconnect();
                    }
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }

            runOnUiThread(this::loaded);
        });

        t.start();
    }

    private void loaded() {
        Intent intent = new Intent(this, AwaitStartActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/space/itoncek/composetest/hideandseek/LoginActivity.kt
package space.itoncek.composetest.hideandseek

import android.app.Activity
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.wrapContentHeight
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.tooling.preview.Wallpapers
import androidx.compose.ui.unit.dp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner
import org.apache.commons.codec.digest.DigestUtils
import space.itoncek.composetest.api.HideAndSeekApiFactory
import space.itoncek.composetest.runOnUiThread
import space.itoncek.composetest.ui.theme.ComposeTestTheme
import space.itoncek.composetest.ui.theme.DesignBg
import space.itoncek.composetest.ui.theme.DesignBlueWhite
import space.itoncek.composetest.ui.theme.DesignFg
import space.itoncek.composetest.ui.theme.DesignShadow
import space.itoncek.trailcompass.client.api.HideAndSeekAPI
import space.itoncek.trailcompass.client.api.LoginResponse
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import kotlin.concurrent.thread

class LoginActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //enableEdgeToEdge()
        setContent {
            ComposeTestTheme {
                LoginView()
            }
        }
    }
}


@Composable
fun LoginView() {
    var progress by remember { mutableFloatStateOf(.10f) }
    var label by remember { mutableStateOf("Loading...") }
    Column(
        modifier = Modifier
            .fillMaxSize(1f)
            .background(MaterialTheme.colorScheme.background)
            .padding(8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Card(
            modifier = Modifier
                .fillMaxWidth(.9f)
                .wrapContentHeight()
                .padding(8.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceContainerHighest),
            elevation = CardDefaults.elevatedCardElevation(8.dp)
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth(1f)
                    .wrapContentHeight()
                    .padding(8.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center
            ) {
                CircularProgressIndicator(
                    progress = { progress },
                    strokeCap = StrokeCap.Round,
                    modifier = Modifier.padding(8.dp, 4.dp),
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    trackColor = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Text(text = label, modifier = Modifier.padding(8.dp, 4.dp), color = MaterialTheme.colorScheme.onSurface)
            }
        }
    }
    val lifecycleOwner = LocalLifecycleOwner.current
    val ctx = LocalContext.current;
    val api: HideAndSeekAPI = HideAndSeekApiFactory(ctx.filesDir).generateApi()

    DisposableEffect(LocalContext.current) {
        val observer = LifecycleEventObserver { source, event ->
            if (event == Lifecycle.Event.ON_START) {
                thread {
                    val cfg = api.config
                    runOnUiThread { label = "Logging in as " + cfg.username }

                    val login = api.login()

                    val activity = ctx as? Activity
                    when (login) {
                        LoginResponse.UNABLE_TO_CONNECT, null -> {
                            runOnUiThread {
                                label = "Unable to connect!"
                                Toast.makeText(ctx, "Unable to connect!", Toast.LENGTH_SHORT)
                                    .show();
                                activity?.finish()
                                return@runOnUiThread
                            }
                        }

                        LoginResponse.INCOMPATIBLE_BACKEND -> {
                            runOnUiThread {
                                label = "Server is on the wrong version!"
                                Toast.makeText(
                                    ctx, "Server is on the wrong version!", Toast.LENGTH_SHORT
                                ).show();
                                activity?.finish()
                                return@runOnUiThread
                            }
                        }

                        LoginResponse.UNABLE_TO_AUTH -> {
                            runOnUiThread {
                                label = "Unable to auth!"
                                Toast.makeText(ctx, "Unable to auth!", Toast.LENGTH_SHORT).show();
                                activity?.finish()

                                return@runOnUiThread
                            }
                        }

                        LoginResponse.OK -> {
                            runOnUiThread {
                                label = "Logged in as ${cfg.username}!"
                                progress = .6f;
                            }
                        }
                    }

                    runOnUiThread {
                        label = "Updating hash"
                        progress = .7f
                    }
                    val mapfile = File(ctx.cacheDir.path + "/map.map");
                    val remoteMapHash = api.mapHash
                    var localMapHash = "";

                    if (mapfile.exists()) {
                        localMapHash = DigestUtils.sha256Hex(FileInputStream(mapfile))
                    }
                    runOnUiThread {
                        label = "Downloading"
                        progress = 1f
                    }
                    if (!remoteMapHash.equals(localMapHash)) {
                        val url = URL(cfg.base_url + "/mapserver/getServerMap")
                        val connection = url.openConnection() as HttpURLConnection
                        connection.setRequestProperty("Authorization", "Bearer " + cfg.jwt_token)
                        connection.connect()

                        // expect HTTP 200 OK, so we don't mistakenly save error report
                        // instead of the file
                        if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                            throw RuntimeException(
                                IOException(
                                    ("Server returned HTTP " + connection.getResponseCode() + " " + connection.getResponseMessage())
                                )
                            )
                        }

                        // this will be useful to display download percentage
                        // might be -1: server did not report the length
                        val fileLength = connection.contentLength
                        Log.i("LoginActivity", "FL $fileLength")

                        // download the file
                        val input = connection.inputStream
                        val output = FileOutputStream(mapfile)

                        val data = ByteArray(65535)
                        var total: Long = 0
                        var count: Int
                        while ((input.read(data).also { count = it }) != -1) {
                            output.write(data, 0, count)
                            total += count.toLong()
                            if (fileLength > 0)  // only if total length is known
                            {
                                val finalTotal = total.toInt()
                                runOnUiThread {
                                    progress = finalTotal.toFloat() / fileLength
                                    label =
                                        "Map download ${(100 * (finalTotal.toFloat() / fileLength)).toInt()}%"
                                }
                            }
                        }
                        Log.i("LoginActivity", "total $total")
                    }

                    runOnUiThread {
                        ctx.startActivity(Intent(ctx, AwaitStartActivity::class.java))
                        val a = ctx as? Activity
                        a?.finish();
                    }
                }
            }
        }
        lifecycleOwner.lifecycle.addObserver(observer)

        onDispose {
            lifecycleOwner.lifecycle.removeObserver(observer)
        }
    }
}

@Preview(
    name = "Dynamic Red Dark",
    group = "dark",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.RED_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Green Dark",
    group = "dark",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.GREEN_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Yellow Dark",
    group = "dark",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.YELLOW_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Blue Dark",
    group = "dark",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.BLUE_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Red Light",
    group = "light",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_NO or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.RED_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Green Light",
    group = "light",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_NO or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.GREEN_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Yellow Light",
    group = "light",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_NO or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.YELLOW_DOMINATED_EXAMPLE
)
@Preview(
    name = "Dynamic Blue Light",
    group = "light",
    showBackground = true,
    uiMode = Configuration.UI_MODE_NIGHT_NO or Configuration.UI_MODE_TYPE_NORMAL,
    wallpaper = Wallpapers.BLUE_DOMINATED_EXAMPLE
)
@Preview(showBackground = true, apiLevel = 33)
@Composable
fun GreetingPreview3() {
    ComposeTestTheme {
        LoginView()
    }
}